pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java

```diff
@@ -597,6 +596,39 @@ public static String splitPart(String input, String delimiter, int index) {
     }
   }
 
+  /**
+   * @param input the input String to be split into parts.
+   * @param delimiter the specified delimiter to split the input string.
+   * @param index the specified index for the splitted parts to be returned.
+   * @param max the max count of parts that the input string can be splitted into.
+   * @return splits string on the delimiter with the limit count and returns String at specified index from the split.
+   */
+  @ScalarFunction
+  public static String splitPart(String input, String delimiter, int index, int max) {
+    String[] splitString = StringUtils.splitByWholeSeparator(input, delimiter, max);
+    if (index < splitString.length) {
+      return splitString[index];
+    } else {
+      return "null";
+    }
+  }
+
+  /**
+   * @param input the input String to be split into parts.
+   * @param delimiter the specified delimiter to split the input string.
+   * @param index the specified index for the splitted parts to be returned.
+   * @return splits string on the delimiter with the limit count and returns String at specified index from the split.
+   */
+  @ScalarFunction
+  public static String splitPartFromEnd(String input, String delimiter, int index) {
```

Comment:
Seems we are already not following the PostgreSQL semantic where `position` starts from 1. So I'd suggest allowing negative value for `index` which indicates the index from end. When it is negative, we can simply pick the one from `length + index`

Reply:
i see. that's a good idea. let me take a look, thanks

Reply:
updated, thanks @Jackie-Jiang can you review again.

Reply:
@Jackie-Jiang, after i adopt to your suggestion, this is a backward incompatible behavior since previously the negative value index will error out (out of range). now these values will indicates the index from end.